File: C/4_vong_lap/bai_22.c
/*
Create on Tue Oct 07 2025 15:12
@author: ntdai

#23 TIM UOC SO CHUNG LON NHAT
	Cach 1: giai thuat Euclid
	Cach 2: su dung phep tru
	Cach 3: su dung phep chia du
*/

// cach 1
#include "stdio.h"
int main(){
	// khai bao 
	int a, b, r;
	// nhap du lieu
	printf("nhap a = ");
	scanf("%d", &a);
	printf("nhap b = ");
	scanf("%d", &b);
	// xu ly
	while(b!=0){
		r = a % b;
		a = b;
		b = r;
	}
	printf("UCLN = %d", a);
}

// cach 2
#include "stdio.h"
int main(){
	// nhap du lieu
	int a, b;
	printf("nhap a = ");
	scanf("%d", &a);
	printf("nhap b = ");
	scanf("%d", &b);
	// xu ly
	if(a==0 || b==0){
		printf("UCLN = %d", a+b);
	}while(a!=b){
		if(a>b){
			a=a-b;	
		}else{
			b=b-a;
		}
	}
	printf("UCLN = %d", a);
}

// cach 3
#include "stdio.h"
int main(){
	// nhap du lieu
	int a, b;
	printf("nhap a = ");
	scanf("%d", &a);
	printf("nhap b = ");
	scanf("%d", &b);
	// xu ly
	while(a*b != 0){
		if(a>b){
			a=a%b;
		}else{
			b=b%a;
		}
	}
	printf("UCLN = %d", a+b);
}

File: python/part_3/bai_22.py
'''
Create Tue Oct 14 2025 15:40
@author: ntdai

#22 BÀI TẬP CHUYỂN ĐỔI SỐ THẬP PHÂN SANG SỐ NHỊ PHÂN 
'''
# nhập n
n = -1
while(n<=0):
    n = int(input("nhập n: "))
#-----------------------------
# chuyển từ thập phân sang nhị phân
binary = ""
if n==0:
    binary = "0"
else:
    while n>0:
        binary = str(n%2) + binary
        n = n // 2
print("nhị phân", binary) 

'''
Ví dụ:
| Lần lặp |  n  | n%2 | old binary |   new binary           | 
|    1    |  13 |  1  |     ""     |   1                    | 
|    2    |  6  |  0  |    "1"     |   "0"+"1" -> "01"      | 
|    3    |  3  |  1  |    "01"    |   "1"+"01" -> "101"    | 
|    4    |  1  |  1  |    "101"   |   "1"+"101" -> "1101"  | 
'''  